flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1420,6 +1424,41 @@ private void timeoutSlot(AllocationID allocationId, UUID ticket) {
 		}
 	}
 
+	/**
+	 * Sync slot with job master.
+	 * The slots only exist in job master will be failed to job master.
+	 * The slots only exist in task executor will be released to resource manager.
+	 *
+	 * @param slotSnapshot is the snapshot of slots in job master
+	 */
+	private void syncSlotWithSnapshotFromJobMaster(SlotSnapshot slotSnapshot) {
+		JobManagerConnection jobManagerConnection = jobManagerTable.get(slotSnapshot.getJobId());
+		if (jobManagerConnection != null) {
+			JobMasterGateway jobMasterGateway = jobManagerConnection.getJobManagerGateway();
+
+			for (OfferedSlot offeredSlot : slotSnapshot.getSlots()) {
+				AllocationID allocationId = offeredSlot.getAllocationId();
+				if (!taskSlotTable.isAllocated(offeredSlot.getSlotIndex(), slotSnapshot.getJobId(), allocationId)) {
+					jobMasterGateway.failSlot(
+							getResourceID(),
+							allocationId,
+							new Exception("Slot " + offeredSlot.getSlotIndex() + " is not allocated by " + allocationId));
```

Comment:
`new FlinkException("Slot " + offeredSlot.getSlotIndex() + " on TaskExecutor " + getResourceID() + " is not allocated by job " + slotSnapshot.getJobId() + '.');`

Reply:
Done, I use String.format() to make it more clear.